Sources/StreamChatUI/Utils/ViewPaginationHandling/StatefulScrollViewPaginationHandler.swift

```diff
@@ -0,0 +1,75 @@
+//
+// Copyright © 2023 Stream.io Inc. All rights reserved.
+//
+
+import UIKit
+
+/// A component responsible to handle when to load new pages in a scrollView holding state associated to the content view
+final class StatefulScrollViewPaginationHandler: StatefulViewPaginationHandling {
+    private var bottomPageRequestItemCount: Int?
+    private var topPageRequestItemCount: Int?
+
+    var canRequestNewTopPage: Bool {
+        topPageRequestItemCount == nil
+    }
+
+    var canRequestNewBottomPage: Bool {
+        bottomPageRequestItemCount == nil
+    }
+
+    private weak var scrollView: UIScrollView?
+    private var observation: NSKeyValueObservation?
+    private var previousPosition: CGFloat = 0.0
+
+    var topThreshold: Int = 100
```

Comment:
it's not clear to me if those values are always expected to be accessed from the same thread. If that's not the case, is it worth protect them behind a serial access?

Reply:
This is internal only, and the expectation is that it is accessed always during the set up/layout of the view, so would not complicate much here